File: 146-lru-cache/lru-cache.java
class LRUCache {

    Map<Integer,Node> track = new HashMap<>();
    int totalCapacity;
    DLL dl = new DLL();

    public LRUCache(int capacity) {
        totalCapacity = capacity;
    }
    
    public int get(int key) {
        if(!track.containsKey(key)) {
                return -1;
            }
        int value = track.get(key).data;
        Node pushed = dl.pushNode(track.get(key));
        track.put(key,pushed);
        return value;
    }
    
    public void put(int key, int value) {
        if(track.containsKey(key)) {
            Node updated = dl.updateNode(track.get(key),value);
            track.put(key,updated);
        } else {
            if (track.size() == totalCapacity) {
                int removedKey = dl.evictKey();
                track.remove(removedKey);
            }
                Node na = dl.addNode(key, value);
                track.put(key, na);
        }
    }

}

class DLL {
    private Node head;
    private Node tail;

    public Node getHead() {
        return head;
    }

    public Node getTail() {
        return tail;
    }

    public Node addNode(int key, int value) {
        if(head == null) {
            head = new Node(key,value);
            tail = head;
            return head;
        }

        Node newNode = new Node(key,value);
        tail.next = newNode;
        newNode.prev = tail;
        tail = newNode;
        return newNode;
    }

    public Node updateNode(Node node,int value) {
        node.data = value;
        node = pushNode(node);
        return node;
    }

    public Node pushNode(Node node) {
        if(head == tail || node.next == null) {
            return node;
        }
        Node pP = node.prev;
        Node nP = node.next;
        if(pP == null) {
            head = head.next;
            head.prev = null;
        } else {
            pP.next = nP;
            nP.prev = pP;
        }
        tail.next = node;
        node.prev = tail;
        node.next = null;
        tail = node;

        return node;
    }

    public int evictKey() {
        if(head == null) return -1;
        int removed;
        if(head.next == null) {
            removed = head.key;
            head = null;
            tail = null;
        } else {
            removed = head.key;
            head = head.next;
            head.prev = null;
        }
        return removed;
    }


}

class Node {
    Node next;
    Node prev;
    int key;
    int data;

    public Node(int key, int value) {
        this.key = key;
        this.data = value;
    }
}






/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: 146-lru-cache/lru-cache.py
class ListNode:
    def __init__(self,key,value):
        self.key = key
        self.value = value
        self.prev = None
        self.next = None

class LRUCache(object):
    def __init__(self, capacity):
        self.capacity = capacity
        self.map = {}
        self.head = ListNode(-1,-1)
        self.tail = ListNode(-1,-1)
        
    def get(self, key):
        if key not in self.map:
            return -1
        else:
            current_node = self.map[key]
            value = current_node.value
            self.moveToTail(key,current_node)
            return value
        
    def put(self, key, value):
        if key not in self.map:
            if len(self.map) == self.capacity:
                key_to_remove = self.removeNode()
                x = self.map
                del self.map[key_to_remove]
            self.addNode(key,value)
        else:
            current_node = self.map[key]
            current_node.value = value
            self.moveToTail(key,current_node)

    def addNode(self,key,value):
        node = ListNode(key,value)
        if self.head.key == -1:
            self.head = node
            self.tail = node
        else:
            node.prev = self.tail
            self.tail.next = node
            self.tail = node
        self.map[key] = node
    
    def removeNode(self):
        removed_key = self.head.key
        if self.head.next is None:
            self.head = ListNode(-1,-1)
            self.tail = ListNode(-1,-1)
        else:
            self.head = self.head.next
            self.head.prev = None
        return removed_key

    def moveToTail(self,key,node):
        if node.next is None or self.head == self.tail:
            return node

        prev_node = node.prev
        next_node = node.next

        if prev_node is None:
            new_head = self.head.next
            new_head.prev = None
            self.head = new_head
        else:
            prev_node.next = next_node
            if next_node:
                next_node.prev = prev_node

        self.tail.next = node
        node.prev = self.tail
        node.next = None
        self.tail = node
        self.map[key] = node

        
# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)